### ISSUE TITLE
LSTM BTC

### ISSUE THREAD

**Issue by i492095-beep** [i492095-beep/Ismail-Newton#1]
# --- YouTube Videonuz İçin Hazır Kod ---
# Gerekli kütüphaneleri yükleyin (izleyicilerinize bunu söyleyin)
!pip install tensorflow pandas numpy yfinance scikit-learn matplotlib

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split

import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense, Dropout

# Veri çekmek için
import yfinance as yf

# --- ADIM 1: Veri Toplama ---
# Yahoo Finance'ten Bitcoin verisini çekelim
# Videonuzda "yatırım tavsiyesi değildir" (YTD) demeyi unutmayın!
data = yf.download('BTC-USD', start='2018-01-01', end='2025-11-16')

# Sadece 'Kapanış' (Close) fiyatına ihtiyacımız var
df = data[['Close']]

# --- ADIM 2: Veri Ön İşleme (Scaling) ---
# Sinir ağları 0-1 arasındaki verilerle daha iyi çalışır
scaler = MinMaxScaler(feature_range=(0, 1))
scaled_data = scaler.fit_transform(df)

# --- ADIM 3: Veri Setini "Pencereleme" (Windowing) ---
# Bu, LSTM için en kritik adımdır.
# Modeli "Geçmiş 60 güne bak, bir sonraki günü tahmin et" diye eğiteceğiz.
window_size = 60
X_data = []
y_data = []

for i in range(window_size, len(scaled_data)):
    # Geçmiş 'window_size' kadar günü (X) al
    X_data.append(scaled_data[i-window_size:i, 0])

    # Bir sonraki günü (y) al
    y_data.append(scaled_data[i, 0])

# NumPy dizilerine dönüştür
X_data, y_data = np.array(X_data), np.array(y_data)

# LSTM Girdisi: [Örnek Sayısı, Zaman Adımı (Pencere), Özellik Sayısı]
# Keras'ın LSTM katmanı bu 3 boyutlu şekli bekler
X_data = np.reshape(X_data, (X_data.shape[0], X_data.shape[1], 1))

# --- ADIM 4: Eğitim ve Test Verisi Ayırma ---
# ÖNEMLİ: Zaman serisi verilerinde veriyi KARIŞTIRAMAYIZ (shuffle=False).
# Geleceği kullanarak geçmişi tahmin edemeyiz. Kronolojik bölmeliyiz.
split_ratio = 0.8
split_index = int(len(X_data) * split_ratio)

X_train, X_test = X_data[:split_index], X_data[split_index:]
y_train, y_test = y_data[:split_index], y_data[split_index:]

print(f"Eğitim Verisi Şekli: {X_train.shape}")
print(f"Test Verisi Şekli: {X_test.shape}")

# --- ADIM 5: LSTM Modelini Oluşturma ---
model = Sequential()

# İlk LSTM katmanı (units=50 nöron)
# return_sequences=True: Bir sonraki LSTM katmanına veri göndereceği için
# input_shape: (60 gün, 1 özellik)
model.add(LSTM(units=50, return_sequences=True, input_shape=(window_size, 1)))
model.add(Dropout(0.2)) # Aşırı öğrenmeyi (overfitting) engellemek için

# İkinci LSTM katmanı
model.add(LSTM(units=50, return_sequences=False)) # Artık çıktı verecek
model.add(Dropout(0.2))

# Çıkış katmanı
# Bir sonraki günün fiyatını (1 nöron) tahmin edeceğiz
model.add(Dense(units=1))

# --- ADIM 6: Modeli Derleme ve Eğitme ---
# Optimizasyon: 'adam' (Gradyan İnişi'nin gelişmiş bir versiyonu)
# Kayıp Fonksiyonu: 'mean_squared_error' (Regresyon için standart)
model.compile(optimizer='adam', loss='mean_squared_error')

# Modeli eğitelim
# epochs=25: Veri seti üzerinden 25 kez geç
# batch_size=32: Her adımda 32'lik gruplar halinde öğren
history = model.fit(X_train, y_train, epochs=25, batch_size=32, validation_data=(X_test, y_test))

# --- ADIM 7: Tahmin Yapma ve Değerlendirme ---
# Test verisi üzerinde tahmin yap
predictions = model.predict(X_test)

# Tahminleri geri ölçeklendir (0-1 arasından gerçek $ değerine)
predictions = scaler.inverse_transform(predictions)

# Gerçek test değerlerini de geri ölçeklendir
y_test_real = scaler.inverse_transform(y_test.reshape(-1, 1))

# --- ADIM 8: Sonuçları Görselleştirme (Videonun Finali!) ---
plt.figure(figsize=(14, 6))
plt.title('Bitcoin Fiyat Tahmini - LSTM Modeli (YTD)')
plt.xlabel('Zaman (Gün)')
plt.ylabel('Fiyat (USD)')

# Gerçek veriyi çiz
plt.plot(y_test_real, color='blue', label='Gerçek Fiyatlar')

# Modelin tahminlerini çiz
plt.plot(predictions, color='red', linestyle='--', label='Tahmin Edilen Fiyatlar')

plt.legend()
plt.show()

# Eğitim kaybını (loss) da gösterebilirsiniz
plt.figure(figsize=(10, 5))
plt.title('Model Eğitim Kaybı (Loss)')
plt.plot(history.history['loss'], label='Eğitim Kaybı')
plt.plot(history.history['val_loss'], label='Test (Validation) Kaybı')
plt.xlabel('Epoch')
plt.ylabel('Loss')
plt.legend()
plt.show()

print("--- Tahmin Başarıyla Tamamlandı ---")

# --- YARININ FİYATINI TAHMİN ETME KODU ---

# 1. "Soru"yu hazırla: En son 'window_size' (60) günlük ölçekli veriyi al
last_60_days_scaled = scaled_data[-window_size:]

# 2. Soruyu modelin anlayacağı 3D formata (1, 60, 1) getir
X_predict = np.reshape(last_60_days_scaled, (1, window_size, 1))

# 3. "Soruyu sor" (Tahmini yap):
# Model, bu 60 güne bakarak 61. gün (yarın) için bir cevap (scaled) üretecek
predicted_price_scaled = model.predict(X_predict)

# 4. "Cevabı" bizim anlayacağımız USD formatına çevir
predicted_price = scaler.inverse_transform(predicted_price_scaled)

# --- Sonuçları Gösterelim ---

# Karşılaştırmak için son bilinen fiyatı (bugünün fiyatı) da alalım
last_real_price = scaler.inverse_transform(scaled_data[-1].reshape(-1, 1))

print("--- 🚀 Gelecek Tahmini ---")
print(f"Bilinen Son Fiyat (Bugün): {last_real_price[0][0]:.2f} $")
print(f"Modelin Tahmini (Yarın): {predicted_price[0][0]:.2f} $")

# Basit bir yön tahmini
if predicted_price > last_real_price:
    print("Modelin Yön Tahmini: YÜKSELECEK 📈")
else:
    print("Modelin Yön Tahmini: DÜŞECEK 📉")
